File: oauth2library/src/main/java/ca/mimic/oauth2library/Access.java
package ca.mimic.oauth2library;

import java.io.IOException;

import okhttp3.Authenticator;
import okhttp3.FormBody;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;

public class Access {
    protected static OAuthResponse refreshAccessToken(String token, OAuth2Client oAuth2Client) throws IOException {
        FormBody.Builder formBodyBuilder = new FormBody.Builder()
                .add(Constants.POST_REFRESH_TOKEN, token);

        Utils.postAddIfValid(formBodyBuilder, oAuth2Client.getFieldsAsMap());

        final Request request = new Request.Builder()
                .url(oAuth2Client.getSite())
                .post(formBodyBuilder.build())
                .build();

        return refreshTokenFromResponse(oAuth2Client, request);
    }

    private static OAuthResponse refreshTokenFromResponse(final OAuth2Client oAuth2Client,
                                                  final Request request) throws IOException {
        return getTokenFromResponse(oAuth2Client, oAuth2Client.getOkHttpClient(),
                request, new AuthState(AuthState.REFRESH_TOKEN));

    }

    protected static OAuthResponse getToken(OAuth2Client oAuth2Client) throws IOException {
        FormBody.Builder formBodyBuilder = new FormBody.Builder();
        Utils.postAddIfValid(formBodyBuilder, oAuth2Client.getFieldsAsMap());
        Utils.postAddIfValid(formBodyBuilder, oAuth2Client.getParameters());

        return getAccessToken(oAuth2Client, formBodyBuilder);
    }

    private static OAuthResponse getAccessToken(OAuth2Client oAuth2Client, FormBody.Builder formBodyBuilder)
            throws IOException {
        final Request request = new Request.Builder()
                .url(oAuth2Client.getSite())
                .post(formBodyBuilder.build())
                .build();

        return getTokenFromResponse(oAuth2Client, request);
    }

    private static OAuthResponse getTokenFromResponse(final OAuth2Client oAuth2Client,
                                              final Request request) throws IOException {
        return getTokenFromResponse(oAuth2Client, oAuth2Client.getOkHttpClient(),
                request, new AuthState(AuthState.ACCESS_TOKEN));

    }

    private static OAuthResponse getTokenFromResponse(final OAuth2Client oAuth2Client,
                                              final OkHttpClient okHttpClient,
                                              final Request request,
                                              final AuthState authState) throws IOException {

        Authenticator authenticator = oAuth2Client.getAuthenticator() == null ? Utils.getAuthenticator(oAuth2Client, authState) : oAuth2Client.getAuthenticator();
        Response response = okHttpClient.newBuilder().authenticator(authenticator)
                .build().newCall(request).execute();

        return new OAuthResponse(response);
    }

}

File: oauth2library/src/main/java/com/pillohealth/oauth2library/Access.kt
package com.pillohealth.oauth2library

import okhttp3.FormBody
import okhttp3.OkHttpClient
import okhttp3.Request

internal object Access {

    internal fun refreshAccessToken(token: String, oAuth2PasswordClient: OAuth2Password2Client): OAuthResponse {
        val formBodyBuilder = FormBody.Builder()
                .add(Constants.POST_REFRESH_TOKEN, token)

        Utils.postAddIfValid(formBodyBuilder, oAuth2PasswordClient.fieldsAsMap)

        val request = Request.Builder()
                .url(oAuth2PasswordClient.site)
                .post(formBodyBuilder.build())
                .build()

        return refreshTokenFromResponse(oAuth2PasswordClient, request)
    }

    private fun refreshTokenFromResponse(oAuth2PasswordClient: OAuth2Password2Client,
                                         request: Request): OAuthResponse {
        return getTokenFromResponse(oAuth2PasswordClient, oAuth2PasswordClient.getOkHttpClient(),
                request, AuthState(AuthState.REFRESH_TOKEN))

    }

    internal fun getToken(oAuth2PasswordClient: OAuth2Password2Client): OAuthResponse {
        val formBodyBuilder = FormBody.Builder()
        Utils.postAddIfValid(formBodyBuilder, oAuth2PasswordClient.fieldsAsMap)
        Utils.postAddIfValid(formBodyBuilder, oAuth2PasswordClient.parameters)

        return getAccessToken(oAuth2PasswordClient, formBodyBuilder)
    }

    private fun getAccessToken(oAuth2PasswordClient: OAuth2Password2Client, formBodyBuilder: FormBody.Builder): OAuthResponse {
        val request = Request.Builder()
                .url(oAuth2PasswordClient.site)
                .post(formBodyBuilder.build())
                .build()

        return getTokenFromResponse(oAuth2PasswordClient, request)
    }

    private fun getTokenFromResponse(oAuth2PasswordClient: OAuth2Password2Client,
                                     request: Request): OAuthResponse {
        return getTokenFromResponse(oAuth2PasswordClient, oAuth2PasswordClient.getOkHttpClient(),
                request, AuthState(AuthState.ACCESS_TOKEN))

    }

    private fun getTokenFromResponse(oAuth2PasswordClient: OAuth2Password2Client,
                                     okHttpClient: OkHttpClient,
                                     request: Request,
                                     authState: AuthState): OAuthResponse {
        val response = okHttpClient.newBuilder().authenticator(Utils.getAuthenticator(oAuth2PasswordClient, authState))
                .build().newCall(request).execute()

        return OAuthResponse(response)
    }

}
